Hide table layouts from engine

martint opened the pull request:
In preparation for https://github.com/prestosql/presto/wiki/Pushdown-of-complex-operations, this change removes support for multiple layouts. It's a feature that's not used by any known connector, complicates the work of the optimizer and just gets in the way. In the future, we may add this functionality again but in a different form.

martint committed:
Remove support for multiple table layouts

martint committed:
Hide TableLayouts from engine

This change hides table layouts from the engine as a first-class
concept. We keep the SPI as is for backward compatibility for now.

When predicates are pushed into a table scan by PickLayout (now
PushPredicateIntoTableScan) or AddExchanges, we now replace the
table handle associated with the table scan with a new one that
contains the reference to the ConnectorTableLayoutHandle under
the covers.

martint committed:
Simplify unconditional PickLayout

It's just selecting a layout for the raw table scan, so
no need to go through the logic for pushing a predicate,
etc.

martint committed:
Inline unnecessary method

martint committed:
Remove unused parameter

martint committed:
Move test to distributed query tests

This test relies on a layout being picked during planning. For
LocalQueryRunner this only happens because of a synthetic rule
(PickLayout w/o predicate) that attaches a layout to the table scan.
So, in a sense, it's just a coincidence that it works.

In distributed execution, the job is done by AddExchanges, so
we want to make sure we're testing that behavior.

martint committed:
Rename listTableLayouts method

More accurately, constructs a set of alternate plans with the
filter pushed into the table scan based on available table layouts.

dain commented on presto-main/src/main/java/io/prestosql/sql/planner/plan/TableScanNode.java:
```diff
@@ -42,48 +40,43 @@
 
     private final TupleDomain<ColumnHandle> enforcedConstraint;
 
+    public static TableScanNode newInstance(
```
This is a strange design, which I think predates this change.  Maybe add a comment on why this factory is different from the constructor with the same signature.

martint replied:
Ah, thanks for pointing it out. I need to see if I can get rid of it. Without it, two of the constructors had the same shape (but ended up doing different things). I may be able to consolidate them now that we don’t have table layouts.

martint replied:
It turns out I can't. I tried making the static factory method the one that's used to deserialized from json, but it needs to set some fields to null, which is disallowed by the full constructor. Oh well.. I added a comment to clarify.

kokosing replied:
But we used to have such constructor and it used to work. What was wrong with it?

dain reviewed (approved):
One comment, but otherwise looks good.

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/TableHandle.java:
```diff
@@ -17,23 +17,31 @@
     private final ConnectorId connectorId;
     private final ConnectorTableHandle connectorHandle;
+    private final ConnectorTransactionHandle transaction;
+    private final Optional<ConnectorTableLayoutHandle> layout;
```
To me it looks like to much of information to be stored in `TableHandle`. To me `TableHandle` represents an handle to a table, same way like a column handle. It is expected to be lightweight. What you have here is designed for very specific use case, hence maybe a new entity should be created for that.

Notice that you are planning to put even more information here (aggregations, joins, projections), the more you add here the less "table"-like it becomes, instead it becomes a more generic "relation" (no longer a table). Also notice that `TableHandle` is used for insert (perfect fit usage) and now with your changes it becomes questionable.

martint replied:
> To me it looks like to much of information to be stored in TableHandle. To me TableHandle represents an handle to a table, same way like a column handle. It is expected to be lightweight. What you have here is designed for very specific use case, hence maybe a new entity should be created for that.

Storing the TableLayout is only temporary. I plan to eventually get rid of them entirely, but didn't do so yet because it requires backward incompatible changes to the SPI. 

But at the end of the day, yes, the plan is for TableHandle to potentially represent more than raw tables. The object doesn't need to be heavyweight -- it's up to the connector to decide what it holds on to internally (it could be just an id to an stored definition, etc). As to whether it becomes more "relation"-like, that's not much of a concern. In fact, the SQL spec treats everything as some form of a table:

```
A table is a collection of zero or more rows where each row is a sequence of one or more column values. 

[...] 

A table is either a base table, a derived table, or a transient table.

[...]

A derived table is a table derived directly or indirectly from one or more other tables by the evaluation of an expression, 
such as a <joined table>, <data change delta table>, <query expression>, or <table expression>. A <query expression> 
can contain an optional <order by clause>
```

> Also notice that TableHandle is used for insert (perfect fit usage) and now with your changes it becomes questionable.

That's ok, because whether a table can be inserted into is decided at analysis time. So you'd never end up with a TableHandle to insert into that represents something that can't be inserted into. BTW, did you know the SQL spec allows for inserting into things that are more complex that simple tables? For example, there's an entire section describing "updatable views".

kokosing replied:
I see your point. Anyway, I still think that `TableHandle` was a simple wrapper around `ConnectorTableHandle` and there are still places where it is used as such. Now, they will get a complex type, so they would need to at least check if the layout is empty there.

> Storing the TableLayout is only temporary. I plan to eventually get rid of them entirely, but didn't do so yet because it requires backward incompatible changes to the SPI.

But there will be something else that will replace `TableLayout`. So `TableHandle` won't become any simpler.

> So you'd never end up with a TableHandle to insert into that represents something that can't be inserted into.

Having separate a type would make expressive in the code that such case is impossible (without knowing the actual flow).

> BTW, did you know the SQL spec allows for inserting into things that are more complex that simple tables?

Mind blowing. Anyway, I don't think that `TableHandle` that is suited for read will match that.

kokosing reviewed (commented):
Can you please refresh my memory and explain why `TableLayout*` were introduced in the first place and why it no longer holds?

martint commented:
> Can you please refresh my memory and explain why TableLayout* were introduced in the first place and why it no longer holds?

It's a feature we added a few years ago for a very specific use case at FB to allow connectors to provide multiple physical organizations of a table for the engine to use during optimization. No one else is using it. Since it makes it hard to introduce new features like complex operation pushdown, we're removing it for now. We may reintroduce it later in some other shape (materialized query tables/materialized views, etc).

kokosing commented:
Would you mind to sort commits, that way it will be easier to review.

martint commented:
> Would you mind to sort commits, that way it will be easier to review.

done

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/TableLayoutResult.java:
```diff
@@ -29,15 +29,22 @@
 
 public class TableLayoutResult
 {
+    private final TableHandle newTableHandle;
```
why `newTableHanlde` and not just `tableHandle`?

martint replied:
It's just to indicate that this is the new table handle to be used to refer to the result of pushing the predicates into the table scan. It's named that way for clarity, but it doesn't matter in the medium term since I'm planning to remove this class altogether once the replacement from https://github.com/prestosql/presto/wiki/Pushdown-of-complex-operations is in place and we give connectors some time to migrate.

kokosing commented on presto-main/src/main/java/io/prestosql/metadata/TableLayoutResult.java:
```diff
@@ -29,15 +29,22 @@
-    public TableLayoutResult(TableLayout layout, TupleDomain<ColumnHandle> unenforcedConstraint)
+    public TableLayoutResult(TableHandle newTable, TableLayout layout, TupleDomain<ColumnHandle> unenforcedConstraint)
     {
+        this.newTableHandle = newTable;
```
rnn

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java:
```diff
@@ -90,20 +90,14 @@ public PickTableLayout(Metadata metadata, SqlParser parser)
     {
         return ImmutableSet.of(
                 checkRulesAreFiredBeforeAddExchangesRule(),
```
is this still required? I would replace with comment in `PlanOptimizers`. There is no reason to prevent this class to be invoked later.

martint replied:
Probably not. I'll remove it.

kokosing commented on presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushPredicateIntoTableScan.java:
```diff
@@ -73,13 +73,13 @@
  * chosen by AddExchanges
  */
-public class PickTableLayout
+public class PushPredicateIntoTableScan
```
I think changes in this class should be extracted to separate commit.

martint replied:
Unfortunately, I don't think I can. The change is logically tied together with removing the notion of table layouts from the engine. If I make these changes in a commit before or after the current one, we're left with a build that doesn't work.

kokosing reviewed (commented):
LGTM for all commits but `Hide TableLayouts from engine`

martint commented:
> But we used to have such constructor and it used to work. What was wrong with it?

Before, we had two constructors:

```java
@JsonCreator
public TableScanNode(
        @JsonProperty("id") PlanNodeId id,
        @JsonProperty("table") TableHandle table,
        @JsonProperty("outputSymbols") List<Symbol> outputs,
        @JsonProperty("assignments") Map<Symbol, ColumnHandle> assignments,
        @JsonProperty("layout") Optional<TableLayoutHandle> tableLayout)
```

and 

```java
public TableScanNode(
        PlanNodeId id,
        TableHandle table,
        List<Symbol> outputs,
        Map<Symbol, ColumnHandle> assignments)
```

After removing table layouts, both constructors have the same signature, but do something different.

martint commented:
> I see your point. Anyway, I still think that TableHandle was a simple wrapper around ConnectorTableHandle and there are still places where it is used as such. Now, they will get a complex type, so they would need to at least check if the layout is empty there.

It will still be a simple wrapper around `ConnectorTableHandle` once we finish removing table layouts from the SPI. But we can'd do that quite yet -- we need to 1) add the new APIs and mechanism for predicate pushdown 2) have a grace period for connectors to adapt (or do it automatically, if possible)

> But there will be something else that will replace TableLayout. So TableHandle won't become any simpler.

No, there won't. It will be just TableHandles wrapping ConnectorTableHandles. It's up to connectors to decide what they need to track in a ConnectorTableHandle based on what can be pushed down into the connector.

> Having separate a type would make expressive in the code that such case is impossible (without knowing the actual flow).

That's really hard to do in general. It forces entire parallel hierarchies and APIs to represent similar objects with different constraints (e.g., we'd need separate APIs to resolve a table during analysis for each possible use so we can get a table handle for insert, a table handle for delete, etc.). Also, where do we stop? Would we have one of each expression kind for each SQL type to prevent misuse? It's a fine balance, and I think the current concepts provide good separation between action (expression/plan node type), object (columns, tables, arguments) and attributes (physical or logical properties) without requiring a NxMxR explosion of valid combinations.

martint committed:
Add null checks

martint committed:
Make PushPredicateIntoTableScan top level rule

It now contains a single rule, so no point in
having it return a rule set.

martint committed:
Remove unnecessary expression translations

In order to translate expression to row expressions, the
code was first replacing all symbol references with field
references for the corresponding ordinal inputs.

This is unnecessary, as the translation can be done on demand
as the expression is translated to a row expression.

martint committed:
Remove interpreted page processors

They were only being used in tests. The engine no longer
relies on them for query execution.

martint committed:
Remove unused functions

martint committed:
Fix expression type

The inferred type of the former expression is INTEGER, which
doesn't match the signature of combineHash function call.

martint committed:
Rename analyzeExpressionsWithSymbols method

There's no longer a conflict with analyzeExpressionsWithInputs so
simplify the name

martint committed:
Inline analyzeExpressions method

There's only one caller, so no need for an extra indirection.

sopel39 commented on presto-main/src/main/java/io/prestosql/metadata/Metadata.java:
```diff
@@ -84,14 +84,14 @@
      * the original partitioning handle associated with the provided table layout handle,
      * as promised by {@link #getCommonPartitioning}.
      */
-    TableLayoutHandle makeCompatiblePartitioning(Session session, TableLayoutHandle tableLayoutHandle, PartitioningHandle partitioningHandle);
```
update javadocs?

sopel39 commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -390,30 +398,39 @@ public boolean catalogExists(Session session, String catalogName)
         }
 
-        return Optional.of(new TableLayoutResult(fromConnectorLayout(connectorId, transaction, layouts.get(0).getTableLayout()), layouts.get(0).getUnenforcedConstraint()));
+        ConnectorTableLayout tableLayout = layouts.get(0).getTableLayout();
```
Should we enforce connectors return only single layout?

martint replied:
Good idea. I'll add that.

sopel39 commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -390,30 +398,39 @@ public boolean catalogExists(Session session, String catalogName)
+
+        return handle.getLayout()
+                .map(layout -> new TableLayout(connectorId, handle.getTransaction(), metadata.getTableLayout(connectorSession, layout)))
+                .orElseGet(() -> getLayout(session, handle, Constraint.alwaysTrue(), Optional.empty())
```
It makes me wonder if `TableHandle` shouldn't have basic layout set at planning time so that layout cannot be absent. Then layout could only be improved later on by pushing down predicates.

martint replied:
I went down that path at some point but it was a dead end. It certainly would have made things easier. The problem is that choosing a layout naively (without filter) causes the hive connector to try to enumerate all partitions in the table.

sopel39 replied:
Could we insert some "dummy" (noop) table handle, but I guess it won't be better than just `Optional.empty()`.

martint replied:
Nah, that requires a bunch of special-casing everywhere and doesn't simplify things much.

sopel39 commented on presto-main/src/main/java/io/prestosql/metadata/TableHandle.java:
```diff
@@ -17,23 +17,31 @@
     private final ConnectorId connectorId;
     private final ConnectorTableHandle connectorHandle;
+    private final ConnectorTransactionHandle transaction;
+    private final Optional<ConnectorTableLayoutHandle> layout;
```
Could you add a Javadoc? E.g: that it's used only for table scans and it's temporary. Maybe add a TODO too

sopel39 commented on presto-main/src/main/java/io/prestosql/metadata/TableLayoutResult.java:
```diff
@@ -25,15 +25,22 @@
 
 public class TableLayoutResult
 {
+    private final TableHandle newTableHandle;
```
Why do we need this field since we still have `layout`?

martint replied:
Layout doesn't have a reference to the ConnectorTableHandle that's needed to construct the new TableHandle. I could add that, but it seemed conceptually cleaner to do it this way.

sopel39 commented on presto-main/src/main/java/io/prestosql/split/SplitManager.java:
```diff
@@ -36,10 +40,15 @@
 
+    // This is used to fetch a table layout if the TableHandle doesn't have one set
+    // It's a temporary measure until we get rid of TableLayouts from the SPI
+    private final Metadata metadata;
```
How it can have unset `TableLayout`? Doesn't that indicate error is planning?

martint replied:
LocalQueryRunner doesn't run AddExchanges (https://github.com/prestosql/presto/issues/275), so if there's no filter that can be pushed down there won't be a layout associated with the TableHandle in the TableScan by the time it gets here.

sopel39 replied:
Hm. Could we change `LocalQueryRunner` instead of `SplitManager`? `SplitManager` seems more essential than `LocalQueryRunner` to the engine so it should have more solid design.

Maybe we should choose some layout anyway by the end of planning? E.g:
in `PlanOptimizers` you could have some rule for that when `forceSingleNode==true`. Then we could also have a validator that layout is always present.

martint replied:
That's a much bigger (non-essential) change, but let me see if I can come up with a way to avoid needing Metadata here.

martint replied:
Found a much simpler way to do it that keeps the hack inside LocalQueryRunner (which is already hacky due to the issue I linked above). Take a look: https://github.com/prestosql/presto/pull/363/commits/500156a7515bc73b50f932ba44e04dc804ed8eee

Thanks for prompting me to think harder about a better way to do it :)

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/HashGenerationOptimizer.java:
```diff
@@ -839,7 +839,7 @@ public HashComputationSet withHashComputation(Optional<HashComputation> hashComp
         }
 
-        Expression result = new LongLiteral(String.valueOf(INITIAL_HASH_VALUE));
+        Expression result = new GenericLiteral(StandardTypes.BIGINT, String.valueOf(INITIAL_HASH_VALUE));
```
Would it fail somehow? Is it possible to write test?

martint replied:
It doesn't fail because it's used only in a benchmark that relies on the interpreter. The interpreter "compensates" for this by doing the implicit cast. 

It does fail when trying to compile the expression, which is what I did in a follow commit to remove the interpreted processors.

sopel39 commented on presto-main/src/test/java/io/prestosql/operator/BenchmarkScanFilterAndProjectOperator.java:
```diff
@@ -216,32 +215,25 @@ private RowExpression getFilter(Type type)
-        private RowExpression rowExpression(String expression, Type type)
+        private RowExpression rowExpression(String value)
         {
-            SymbolToInputRewriter symbolToInputRewriter = new SymbolToInputRewriter(sourceLayout);
```
Can we remove `SymbolToInputRewriter` completely

martint replied:
It’s coming. There are a couple of other dependencies, but I have a change in the complex pushdown PR to do it.

sopel39 reviewed (commented):
well done!

martint commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/MetadataQueryOptimizer.java:
```diff
@@ -139,10 +139,9 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
-                if (layouts.size() == 1) {
-                    layout = Iterables.getOnlyElement(layouts).getLayout();
-                }
+                layout = metadata.getLayout(session, tableScan.getTable(), Constraint.alwaysTrue(), Optional.empty())
```
What's wrong with the formatting?

sopel39 replied:
nvm

dain commented on presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java:
```diff
@@ -144,12 +145,14 @@ public static RowExpression translate(
             Session session,
-            boolean optimize)
+            boolean optimize,
+            Map<Symbol, Integer> layout)
```
I'd put this below `type`

dain commented on presto-main/src/test/java/io/prestosql/operator/scalar/FunctionAssertions.java:
```diff
@@ -867,29 +868,45 @@ protected Void visitSymbolReference(SymbolReference node, Void context)
+            }
+        });
+
+        Block block = Utils.nativeValueToBlock(expectedType, result);
```
add comment `// convert result from stack type to Type ObjectValue`

rongrong commented on presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java:
```diff
@@ -252,10 +251,10 @@ public Object evaluate()
     }
 
-    public Object evaluate(int position, Page page)
+    public Object evaluate(SymbolResolver inputs)
```
What's the difference between this `evaluate` and the `optimize` now? If you provide `SymbolResolver`, using `optimize` would be just fine, right?
